Setup Project

CMDRyan opened the pull request:
Initiate required files 

- gitlabAPI.js : Defined to fetch Projects and Jobs using API provided by gitlab.
- gitJobs.js : Get and filter required data from gitlab using provided group ID.
- app.js: Localhost server on port: 8081.
Additionally, empty ejs template engine views are added.

Usage:

- $export API_TOKEN=`your token`
- $node app.js

- Browser: http://localhost:8081/groups/ `group ID` /jobs

Steven-Chan commented on getJobs.js:
```diff
@@ -0,0 +1,53 @@
+  return new Promise((resolve) => {
+
+    const apiToken = process.env.API_TOKEN; 
+    const gitlabAPI = require("./gitlabAPI");
```
Use require at top level in this file.

Steven-Chan commented on getJobs.js:
```diff
@@ -0,0 +1,53 @@
+module.exports.getJobs = function getJobs(requestedGroupID) {
+  return new Promise((resolve) => {
+
+    const apiToken = process.env.API_TOKEN; 
```
Constants should also be put in top level, or you can create a new file to put all server config called `config.js` and put all env var in it.

Steven-Chan commented on getJobs.js:
```diff
@@ -0,0 +1,53 @@
+module.exports.getJobs = function getJobs(requestedGroupID) {
+  return new Promise((resolve) => {
```
You don't need to create a new promise in this way. We only `new Promise` when transforming non-promise functions to promise.

Steven-Chan commented on getJobs.js:
```diff
@@ -0,0 +1,53 @@
+      .getProjectsByGroupID(groupID, apiToken, getProjectSearchParams)
+
+      .then((data) => {
+        const ids = [];
+        for (let i = 0; i < data.length; i += 1) {
+          const item = data[i];
+          ids[i] = item.id;
+        }
+        return ids;
```
Try using `map` of array to simplify the code.

Steven-Chan commented on getJobs.js:
```diff
@@ -0,0 +1,53 @@
+        return ids;
+      })
+      .then((projectIDs) => {
+        const data = [];
+        for (let i = 0; i < projectIDs.length; i += 1) {
+          data[i] = gitlabAPI.getJobsByProjectID(projectIDs[i], apiToken);
+        }
+        return Promise.all(data);
```
Try `map`.

Steven-Chan commented on getJobs.js:
```diff
@@ -0,0 +1,53 @@
+        const sortedData = flattenedJobArray.sort(
+          (a, b) => a.created_at - b.created_at
+        );
+        for (let i = 0, j = 0; i < sortedData.length; i += 1) {
+          const jobStatus = sortedData[i].status;
+
+          if (
+            jobStatus === "success" ||
+            jobStatus === "canceled" ||
+            jobStatus === "failed" ||
+            jobStatus === "skipped"
+          ) {
+          } else {
+            outputJobs[j] = sortedData[i];
+            j++;
+          }
+        }
```
Try `filter`.

Steven-Chan commented on gitlabAPI.js:
```diff
@@ -0,0 +1,32 @@
+const fetch = require("node-fetch");
+
+exports.getProjectsByGroupID = function getProjectsByGroupID(
+  GroupID,
```
Do not use capital letter to start a param variable name.

Steven-Chan commented on gitlabAPI.js:
```diff
@@ -0,0 +1,32 @@
+  });
+};
+
+exports.getJobsByProjectID = function getJobsByProjectID(ProjectID, key) {
```
Do not use capital letter to start a param variable name.

Steven-Chan commented on gitlabAPI.js:
```diff
@@ -0,0 +1,32 @@
+const fetch = require("node-fetch");
+
+exports.getProjectsByGroupID = function getProjectsByGroupID(
+  GroupID,
+  key,
+  searchParam
```
Please organize the function in better way. JS function param is unnamed, so the ordering is important. We group similar things together, so `key` should not be in the middle of `groupID` and `searchParams`.

Steven-Chan commented on package.json:
```diff
@@ -0,0 +1,32 @@
+    "prettier": "prettier --write ."
+  },
+  "dependencies": {
+    "@oursky/eslint-plugin": "^5.0.0",
+    "ejs": "^3.1.6",
+    "eslint": "^7.31.0",
+    "express": "^4.17.1",
+    "node-fetch": "2.6.1",
+    "postcss-import": "^14.0.2",
+    "tailwindcss": "^2.2.4"
+  },
+  "devDependencies": {
+    "@typescript-eslint/eslint-plugin": "^4.28.3",
+    "@typescript-eslint/parser": "^4.28.3",
+    "autoprefixer": "^10.3.1",
+    "eslint-config-airbnb": "^18.2.1",
+    "eslint-config-prettier": "^8.3.0",
+    "eslint-plugin-import": "^2.23.4",
+    "eslint-plugin-jsx-a11y": "^6.4.1",
+    "eslint-plugin-react": "^7.24.0",
+    "eslint-plugin-react-hooks": "^4.2.0",
+    "postcss": "^8.3.5",
+    "prettier": "2.3.2"
```
Please pin all to exact version and run `yarn` again.

Steven-Chan replied:
Not addressed

Steven-Chan commented on views/partials/head.ejs:
```diff
@@ -0,0 +1,9 @@
+<meta charset="UTF-8">
+<title>Testing title</title>
+<!-- CSS (load bootstrap from a CDN) -->
+<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.5.2/css/bootstrap.min.css">
```
We use tailwind instead of bootstrap, so please remove this.

Steven-Chan commented on views/src/styles.css:
```diff
@@ -0,0 +1,133 @@
+@import url("https://fonts.googleapis.com/css2?family=Quicksand&display=swap");
```
This PR does not have UI yet, so please remove this file.

On the other hand, it looks to me that defining style here defeat the purpose of using tailwind.

Steven-Chan reviewed (changes requested):
1. Please remove blackbox related thing in this commit
2. Do not check in tailwind.css here, it is a generated file.
3. Put source code files in `src` folder instead of the top level. Otherwise you will mix build phase code and runtime code together.

Put the steps to setup this project in README instead of PR comment.

Steven-Chan commented:
testing

Steven-Chan commented:
@CMDRyan 

- blackbox not removed yet
- The tailwind.css is needed, but not committed and no instruction about it
- Dont say discussed minor issue in commit, list out what u have changed
- It is impossible to have both package-lock.json and yarn-lock.json at the same time

CMDRyan commented:
@Steven-Chan 

- Resent commit message
- Removed 'blackbox' related lines
- Removed package.json
- Added tailwind.css
-

CMDRyan commented:
@Steven-Chan 

Updated README with line: $yarn add express .

CMDRyan commented:
@Steven-Chan

- Updated README `yarn add express` to `yarn`
- Relocated all source code files to /src and corrected path in app.js
- getJobs.js will now return an array of Job objects
